File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
function solution(maps) {
    const dy = [1,0,-1,0];
    const dx = [0,1,0,-1];
    const n = maps.length;
    const m = maps[0].length;

    const visitCount = [...maps].map((r) => r.map((c) => 1));

    const queue = [[0,0]];

    while(queue.length) {
        const [y, x] = queue.shift();

        for (let i = 0; i < 4; i++) {
            const ny = y + dy[i];
            const nx = x + dx[i];

            if(ny >= 0 && nx >= 0 && ny < n && nx < m) {
                if(maps[ny][nx] === 1 && visitCount[ny][nx] === 1) {
                    visitCount[ny][nx] = visitCount[y][x] + 1;
                    queue.push([ny,nx]);
                }
            }
        }
    }

    return visitCount[n - 1][m - 1] === 1 ? -1 : visitCount[n - 1][m - 1];    
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    move = [[-1, 0], [0, -1], [0, 1], [1, 0]]
    n,m = len(maps),len(maps[0])
    dist = [[-1] * m for _ in range(n)]

    def bfs(s):
        q = deque([s])
        dist[s[0]][s[1]] = 1
        while q:
            cur = q.popleft()
            for direct in move:
                r, c = cur[0] + direct[0], cur[1] + direct[1]
                if r < 0 or r >= n or c < 0 or c >= m: continue
                if maps[r][c] == 0: continue
                if dist[r][c] == -1:
                    q.append([r, c])
                    dist[r][c] = dist[cur[0]][cur[1]] + 1
        return dist

    bfs([0, 0])
    return dist[n - 1][m - 1]